dali/kernels/common/transpose.h

```diff
@@ -0,0 +1,214 @@
+                    span<const int64_t> src_stride, TensorShape<> size, span<const int> perm) {
+  auto dst_level_stride = dst_stride[level];
+  auto src_level_stride = src_stride[perm[level]];
+
+  if (max_levels - level == Static) {
+    transpose_impl_static<T, Static>(dst, src, max_levels, dst_stride, src_stride, size, perm);
+  } else {
+    for (int64_t i = 0; i < size[level]; i++) {
+      transpose_impl<T, Static>(dst, src, level + 1, max_levels, dst_stride, src_stride, size,
+                                perm);
+      dst += dst_level_stride;
+      src += src_level_stride;
+    }
+  }
+}
+
+/**
+ * @brief Find blocks of consecutive numbers in the permutation that would be transposed together
+ *
+ * For example [0, 1, 5, 2, 3, 4] has following blocks: [[0, 1], [5], [2, 3, 4]]
+ * and the result would be: {{0, 1}, {2, 2}, {3, 5}}
+ *
+ * @return List of starting and ending position of blocks(both ends inclusive)
+ */
+std::vector<std::pair<int, int>> PermutationBlocks(span<const int> perm) {
+  if (perm.empty()) {
+    return {};
+  }
+  std::vector<std::pair<int, int>> result;
+  int current_start = 0;
+  int current_end = 0;
+  for (int i = 0; i < perm.size() - 1; i++) {
+    if (perm[i] == perm[i + 1] - 1) {
+      current_end = i + 1;
+    } else {
+      result.push_back({current_start, current_end});
+      current_start = i + 1;
+      current_end = i + 1;
+    }
+  }
+  // We wouldn't add the last one in any case
+  result.emplace_back(current_start, current_end);
+  return result;
+}
+
+/**
+ * @brief Based on the blocks returned from PermutationBlocks, collapse the blocks
+ *        in the perm and re-enumerate them.
+ *
+ * @param perm Permutation to collapse
+ * @param perm_blocks Description of consecutive blocks in perm
+ * @return std::vector<int>
+ */
+std::vector<int> CollapsePermutation(span<const int> perm,
+                                     const std::vector<std::pair<int, int>> &perm_blocks) {
+  std::vector<int> result;
+  std::vector<int> removed;
+  result.reserve(perm_blocks.size());
+  removed.reserve(perm.size() - perm_blocks.size());
+  for (auto &block : perm_blocks) {
+    result.push_back(perm[block.first]);
+    for (int i = block.first + 1; i <= block.second; i++) {
+      removed.push_back(perm[i]);
+    }
+  }
+  for (auto &elem : result) {
+    int original_elem = elem;
+    for (auto removed_elem : removed) {
+      if (original_elem > removed_elem) {
+        elem--;
+      }
+    }
+  }
+  return result;
+}
+
+/**
+ * @brief Based on the blocks returned from PermutationBlocks, collapse the blocks
+ *        in the shape that are permuted together.
+ *
+ * @param shape Shape to be collapsed
+ * @param perm Original, non-collapsed permutation
+ * @param perm_blocks Description of blocks in permutation
+ * @return TensorShape<>
+ */
+TensorShape<> CollapseShape(const TensorShape<> &shape, span<const int> perm,
```

Comment:
This function actually does two things: collapse and permute. The name does not suggest so.
Perhaps it should be called CollapsePermuteShape or PermuteCollapseShape.

Reply:
It used CollapsePermutation for the ease of implementing the shape-collapsing. I removed that call and used different implementation.